crates/block-producer/src/block_builder/prover/block_witness.rs

```diff
@@ -77,13 +48,83 @@ impl BlockWitness {
             .map(|(batch_index, batch)| (batch_index, batch.output_notes_root()))
             .collect();
 
-        Ok(Self {
-            updated_accounts,
-            batch_created_notes_roots,
-            produced_nullifiers: block_inputs.nullifiers,
-            chain_peaks: block_inputs.chain_peaks,
-            prev_header: block_inputs.block_header,
-        })
+        // Order account updates by account ID and each update's initial state hash.
+        //
+        // This let's us chronologically order the updates per account across batches.
+        let mut updated_accounts = BTreeMap::<AccountId, BTreeMap<Digest, AccountUpdate>>::new();
+        for (account_id, update) in batches.iter().flat_map(TransactionBatch::updated_accounts) {
+            updated_accounts
+                .entry(*account_id)
+                .or_default()
+                .insert(update.init_state, update.clone());
+        }
+
+        // Build account witnesses.
+        let mut account_witnesses = Vec::with_capacity(updated_accounts.len());
+        let mut block_updates = Vec::with_capacity(updated_accounts.len());
+
+        for (account_id, mut updates) in updated_accounts {
+            let (initial_state_hash, proof) = block_inputs
+                .accounts
+                .remove(&account_id)
+                .map(|witness| (witness.hash, witness.proof))
+                .ok_or(BuildBlockError::InconsistentAccountIds(vec![account_id]))?;
```

Comment:
We lost some error information because this performs the check for one account instead of all of them. If its important that we catch the entire set of IDs then that is possible, just tedious.